src/frontend/Parser.y

```diff
@@ -769,6 +782,22 @@ checkIterVarType (G.Id tk@(T.Token {T.cleanedString=idName})) = do
         return ()
       else return ()
 
+checkSwitchCaseType :: G.Expr -> ST.ParserMonad ()
+checkSwitchCaseType expr = do
+  ST.SymTable {ST.stSwitchTypes=swTypes} <- RWS.get
+  if swTypes == [] then return ()
+  else do
+    let swType = last swTypes
+    case swType of
+      T.TypeError -> return ()
+      tp -> do
+        let exprType = G.expType expr
+        if exprType `T.canBeConvertedTo` tp then return ()
```

Comment:
if the guard expression can be converted to the switch argument exception we need to wrap the guard expression in a _caster_. The `addCastToExprs` function already does that. You'd need to change the `checkSwitchCaseType` signature to return a `Maybe Expr`